## 2. Cities in Sweden - real dataset


Download the file komtopp50_2020.xlsx from the course github repo. The data is taken from SCB.

&nbsp; a) Read in the tab "Totalt" into a DataFrame and start exploring the data with some simple explorations such as

```python
df.head()
df.info()
df.describe()
```

Feel free to do more explorations. (\*)


In [8]:
import pandas as pd

df = pd.read_excel("../../data/komtopp50_2020.xlsx", sheet_name="Totalt")

df.head()

Unnamed: 0  Unnamed: 1 Unnamed: 2  \
0                                                NaN         NaN        NaN   
1  Folkmängd per kommun 31 december 2020 jämfört ...         NaN        NaN   
2  Sorterat i alfabetisk ordning. Rangordning frå...         NaN        NaN   
3                                                NaN         NaN        NaN   
4                                               Rang         NaN     Kommun   

              Unnamed: 3  Unnamed: 4  Unnamed: 5  
0                    NaN         NaN         NaN  
1                    NaN         NaN         NaN  
2                    NaN         NaN         NaN  
3                    NaN         NaN         NaN  
4  Folkmängd 31 december         NaN  Förändring

&nbsp; b) Clean your data so that the head looks like this: (\*)

|     | Rang 2020 | Rang 2019 | Kommun   | Folkmängd 2020 | Folkmängd 2019 | Förändring |
| --: | --------: | --------: | :------- | -------------: | -------------: | ---------: |
|   0 |        83 |        84 | Ale      |          31868 |          31402 |    1.48398 |
|   1 |        64 |        64 | Alingsås |          41602 |          41420 |   0.439401 |
|   2 |       123 |       123 | Alvesta  |          20224 |          20134 |   0.447005 |
|   3 |       255 |       255 | Aneby    |           6821 |           6848 |  -0.394276 |
|   4 |       169 |       167 | Arboga   |          14039 |


In [ ]:
def clean_data(df):
    df = (
        df.iloc[6:]
        .rename(
            columns={
                "Unnamed: 0": "Rang 2020",
                "Unnamed: 1": "Rang 2019",
                "Unnamed: 2": "Kommun",
                "Unnamed: 3": "Folkmängd 2020",
                "Unnamed: 4": "Folkmängd 2019",
                "Unnamed: 5": "Förandring",
            }
        )
        .sort_values(by="Kommun")
        .reset_index(drop=True)
    )

    df["Rang 2019"] = df["Rang 2019"].astype(int)

    return df


clean_df = clean_data(df)

clean_df.head()

Rang 2020  Rang 2019    Kommun Folkmängd 2020  Folkmängd 2019 Förandring
0        84         85       Ale          15612         15348.0   1.720094
1        64         64  Alingsås          20900         20774.0   0.606527
2       123        123   Alvesta           9857          9795.0   0.632976
3       255        255     Aneby           3318          3350.0  -0.955224
4       165        164    Arboga           6945          6973.0  -0.401549

## 3. Cities in Sweden - gender

We continue with the same Excel-file as in task 2, but now you should also read in the sheets "Kvinnor" and "Män" into two additional DataFrames. In this task, many operations are similar to all three datasets, try creating custom made functions to reuse as much code as possible.

&nbsp; a) Clean your data so that the head looks like this: (\*)

Male:

|     | Rang 2020 | Rang 2019 | Kommun   | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön |
| --: | --------: | --------: | :------- | -------------: | -------------: | ---------: | :-- |
|   0 |        81 |        83 | Ale      |          16256 |          16054 |    1.25825 | Man |
|   1 |        64 |        64 | Alingsås |          20702 |          20646 |   0.271239 | Man |
|   2 |       122 |       123 | Alvesta  |          10367 |          10339 |   0.270819 | Man |
|   3 |       255 |       255 | Aneby    |           3503 |           3498 |   0.142939 | Man |
|   4 |       169 |       168 | Arboga   |           7094 |           7114 |  -0.281136 | Man |

Female:

|     | Rang 2020 | Rang 2019 | Kommun   | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön    |
| --: | --------: | --------: | :------- | -------------: | -------------: | ---------: | :----- |
|   0 |        84 |        85 | Ale      |          15612 |          15348 |    1.72009 | Kvinna |
|   1 |        64 |        64 | Alingsås |          20900 |          20774 |   0.606527 | Kvinna |
|   2 |       123 |       123 | Alvesta  |           9857 |           9795 |   0.632976 | Kvinna |
|   3 |       255 |       255 | Aneby    |           3318 |           3350 |  -0.955224 | Kvinna |
|   4 |       165 |       164 | Arboga   |           6945 |           6973 |  -0.401549 | Kvinna |


In [28]:
male_df = pd.read_excel("../../data/komtopp50_2020.xlsx", sheet_name="Män")
female_df = df = pd.read_excel("../../data/komtopp50_2020.xlsx", sheet_name="Kvinnor")

male_df = clean_data(male_df)
female_df = clean_data(female_df)

male_df["Kön"] = "Man"
female_df["Kön"] = "Kvinna"

female_df.head()

Rang 2020  Rang 2019    Kommun Folkmängd 2020  Folkmängd 2019 Förandring  \
0        84         85       Ale          15612         15348.0   1.720094   
1        64         64  Alingsås          20900         20774.0   0.606527   
2       123        123   Alvesta           9857          9795.0   0.632976   
3       255        255     Aneby           3318          3350.0  -0.955224   
4       165        164    Arboga           6945          6973.0  -0.401549   

      Kön  
0  Kvinna  
1  Kvinna  
2  Kvinna  
3  Kvinna  
4  Kvinna

&nbsp; b) Merge the male and female DataFrames vertically and set index to "Kommun". Note that there now should be 580 rows now. (\*)

|     | Kommun     | Folkmängd 2020 | Folkmängd 2019 | Förändring | Kön    |
| --: | :--------- | -------------: | -------------: | ---------: | :----- |
|   0 | Ale        |          16256 |          16054 |    1.25825 | Man    |
|   1 | Alingsås   |          20702 |          20646 |   0.271239 | Man    |
| ... | ...        |            ... |            ... |        ... | ...    |
| 288 | Överkalix  |           1559 |           1578 |   -1.20406 | Kvinna |
| 289 | Övertorneå |           2027 |           2065 |   -1.84019 | Kvinna |


In [32]:
merged_df = pd.concat([male_df, female_df]).set_index("Kommun")

merged_df

Rang 2020  Rang 2019 Folkmängd 2020  Folkmängd 2019 Förandring  \
Kommun                                                                        
Ale                 81         83          16256         16054.0   1.258253   
Alingsås            64         64          20702         20646.0   0.271239   
Alvesta            122        123          10367         10339.0   0.270819   
Aneby              255        255           3503          3498.0   0.142939   
Arboga             169        168           7094          7114.0  -0.281136   
...                ...        ...            ...             ...        ...   
Österåker           55         57          22997         22442.0   2.473042   
Östhammar          117        118          10846         10803.0   0.398038   
Östra Göinge       160        159           7270          7262.0   0.110162   
Överkalix          284        284           1559          1578.0  -1.204056   
Övertorneå         279        279           2027          2065.0  -1.840194   

                 Kön  
Kommun                
Ale              Man  
Alingsås         Man  
Alvesta          Man  
Aneby            Man  
Arboga           Man  
...              ...  
Österåker     Kvinna  
Östhammar     Kvinna  
Östra Göinge  Kvinna  
Överkalix     Kvinna  
Övertorneå    Kvinna  

[580 rows x 6 columns]